### PR TITLE
initial contribution of Sonos AudioSink support

### PR DESCRIPTION
**Pull request by kaikreuzer**
Note: This PR requires https://github.com/eclipse/smarthome/pull/2305 to be merged first.

implements https://github.com/eclipse/smarthome/issues/1200

### CODE REVIEW

File: extensions/binding/org.eclipse.smarthome.binding.sonos/src/main/java/org/eclipse/smarthome/binding/sonos/internal/SonosHandlerFactory.java [hunk 1]
```diff
@@ -66,23 +73,43 @@ public Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
...
+
+            ZonePlayerHandler handler = new ZonePlayerHandler(thing, upnpIOService, discoveryServiceRegistry, opmlUrl);
+
+            // register the speaker as an audio sink
+            SonosAudioSink audioSink = new SonosAudioSink(handler, audioHTTPServer);
+            @SuppressWarnings("unchecked")
+            ServiceRegistration<AudioSink> reg = (ServiceRegistration<AudioSink>) bundleContext
+                    .registerService(AudioSink.class.getName(), audioSink, new Hashtable<String, Object>());
+            audioSinkRegistrations.put(thing.getUID().toString(), reg);
+
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.sonos/src/main/java/org/eclipse/smarthome/binding/sonos/internal/SonosHandlerFactory.java]: > Trying to understand the reasons why you rather choose the pattern to have SonosAudioSinks in parallel to ZonePlayerHandlers

Very simple: The ZonePlayerHandler class is a huge monolithic monster with 2700 loc, that I didn't want to blow up even further. So I very much preferred to have this functionality separate, especially as there is no need for the handler itself to have any kind of dependency on all the audio stuff.

> Did you already test with a bunch of ZonePlayers in parallel playing the same AudioStream? Evident use case is literally the doorbell.mp3

No, I leave this to you as you have the perfect test setup for it :-)

**Comment by kaikreuzer** [comment]
@maggu2810 Do you want to wait for @kgoderis to test it first or shall we go ahead with review&merge?

**Comment by kgoderis** [comment]
@maggu2810 @kaikreuzer no need to wait for me. I prefer to get my operational env. to the next Stable release including these changes, and then test it in a real world environment instead.

**Comment by maggu2810** [comment]
As I own some Sonos devices myself I can test it, too.
I applied the changes of this PR also for the Chromecast binding and realized that the Karaf runtime does not shutdown "normally" anymore (if you trigger the shutdown you have to wait for minutes until the runtime is closed).
Don't know if this caused by this PR or some other changes in my code base -- need to check that.

And the current code base does not work for me because port (on URI creation) is null on my runtimes. I have created #2331 so ESH don't need to force the product to use a system property to set the port.

**Comment by maggu2810** [comment]
It seems the shutdown hang is not related to this PR.

I added a new Sonos thing (Play 1) and tried to play an audio stream:
```text
2016-10-17 15:14:56,236 | ERROR | nsole user karaf | ConsoleInterpreter               | 147 - org.eclipse.smarthome.io.console - 0.9.0.201610171249 | An error occured while executing the console command.
java.lang.NullPointerException
	at java.math.BigDecimal.<init>(BigDecimal.java:806)[:1.8.0_102]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.getNotificationSoundVolume(ZonePlayerHandler.java:1472)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171249]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.applyNotificationSoundVolume(ZonePlayerHandler.java:2268)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171249]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.handleEmptyQueue(ZonePlayerHandler.java:2255)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171249]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.playNotificationSoundURI(ZonePlayerHandler.java:2140)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171249]
	at org.eclipse.smarthome.binding.sonos.internal.SonosAudioSink.process(SonosAudioSink.java:84)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171249]
	at org.eclipse.smarthome.core.audio.AudioManager.play(AudioManager.java:85)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171249]
	at org.eclipse.smarthome.core.audio.AudioManager.playFile(AudioManager.java:113)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171249]
	at org.eclipse.smarthome.core.audio.internal.AudioConsoleCommandExtension.play(AudioConsoleCommandExtension.java:117)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171249]
	at org.eclipse.smarthome.core.audio.internal.AudioConsoleCommandExtension.execute(AudioConsoleCommandExtension.java:59)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171249]
	at org.eclipse.smarthome.io.console.ConsoleInterpreter.execute(ConsoleInterpreter.java:52)[147:org.eclipse.smarthome.io.console:0.9.0.201610171249]
	at org.eclipse.smarthome.io.console.karaf.internal.CommandWrapper.execute(CommandWrapper.java:58)[148:org.eclipse.smarthome.io.console.karaf:0.9.0.201610171249]
	at org.apache.karaf.shell.impl.console.osgi.secured.SecuredCommand.execute(SecuredCommand.java:67)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.karaf.shell.impl.console.osgi.secured.SecuredCommand.execute(SecuredCommand.java:87)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.executeCmd(Closure.java:480)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.executeStatement(Closure.java:406)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Pipe.run(Pipe.java:108)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.execute(Closure.java:182)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.execute(Closure.java:119)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.CommandSessionImpl.execute(CommandSessionImpl.java:94)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.karaf.shell.impl.console.ConsoleSessionImpl.run(ConsoleSessionImpl.java:274)[80:org.apache.karaf.shell.core:4.0.7]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_102]
```

**Comment by maggu2810** [comment]
Update:
I added the fix of #2323 to #2331 so port and IPv4 should now work more reliable.

After added a Play:1 thing to the system I have to set the notification volume once.
If this has been done, the doorbell is played correctly (triggered by the command line).

Without setting the notification volume manually I run into the NPE (as already written above):
```text
2016-10-17 16:13:18,848 | ERROR | nsole user karaf | ConsoleInterpreter               | 147 - org.eclipse.smarthome.io.console - 0.9.0.201610171358 | An error occured while executing the console command.
java.lang.NullPointerException
	at java.math.BigDecimal.<init>(BigDecimal.java:806)[:1.8.0_102]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.getNotificationSoundVolume(ZonePlayerHandler.java:1472)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.applyNotificationSoundVolume(ZonePlayerHandler.java:2268)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.handleNotificationSound(ZonePlayerHandler.java:2222)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.handleSharedQueue(ZonePlayerHandler.java:2206)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.playNotificationSoundURI(ZonePlayerHandler.java:2138)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.internal.SonosAudioSink.process(SonosAudioSink.java:84)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.AudioManager.play(AudioManager.java:85)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.AudioManager.playFile(AudioManager.java:113)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.internal.AudioConsoleCommandExtension.play(AudioConsoleCommandExtension.java:117)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.internal.AudioConsoleCommandExtension.execute(AudioConsoleCommandExtension.java:59)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.io.console.ConsoleInterpreter.execute(ConsoleInterpreter.java:52)[147:org.eclipse.smarthome.io.console:0.9.0.201610171358]
	at org.eclipse.smarthome.io.console.karaf.internal.CommandWrapper.execute(CommandWrapper.java:58)[148:org.eclipse.smarthome.io.console.karaf:0.9.0.201610171358]
	at org.apache.karaf.shell.impl.console.osgi.secured.SecuredCommand.execute(SecuredCommand.java:67)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.karaf.shell.impl.console.osgi.secured.SecuredCommand.execute(SecuredCommand.java:87)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.executeCmd(Closure.java:480)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.executeStatement(Closure.java:406)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Pipe.run(Pipe.java:108)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.execute(Closure.java:182)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.execute(Closure.java:119)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.CommandSessionImpl.execute(CommandSessionImpl.java:94)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.karaf.shell.impl.console.ConsoleSessionImpl.run(ConsoleSessionImpl.java:274)[80:org.apache.karaf.shell.core:4.0.7]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_102]
```

**Comment by maggu2810** [comment]
> If this has been done, the doorbell is played correctly (triggered by the command line).

It is played correctly, but I still see a NPE in the log.

```text
2016-10-17 16:26:39,715 | ERROR | nsole user karaf | ConsoleInterpreter               | 147 - org.eclipse.smarthome.io.console - 0.9.0.201610171358 | An error occured while executing the console command.
java.lang.NullPointerException
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.waitForFinishedNotification(ZonePlayerHandler.java:2284)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.handleNotificationSound(ZonePlayerHandler.java:2228)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.handleSharedQueue(ZonePlayerHandler.java:2206)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.playNotificationSoundURI(ZonePlayerHandler.java:2138)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.binding.sonos.internal.SonosAudioSink.process(SonosAudioSink.java:84)[213:org.eclipse.smarthome.binding.sonos:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.AudioManager.play(AudioManager.java:85)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.AudioManager.playFile(AudioManager.java:113)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.internal.AudioConsoleCommandExtension.play(AudioConsoleCommandExtension.java:117)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.core.audio.internal.AudioConsoleCommandExtension.execute(AudioConsoleCommandExtension.java:59)[137:org.eclipse.smarthome.core.audio:0.9.0.201610171358]
	at org.eclipse.smarthome.io.console.ConsoleInterpreter.execute(ConsoleInterpreter.java:52)[147:org.eclipse.smarthome.io.console:0.9.0.201610171358]
	at org.eclipse.smarthome.io.console.karaf.internal.CommandWrapper.execute(CommandWrapper.java:58)[148:org.eclipse.smarthome.io.console.karaf:0.9.0.201610171358]
	at org.apache.karaf.shell.impl.console.osgi.secured.SecuredCommand.execute(SecuredCommand.java:67)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.karaf.shell.impl.console.osgi.secured.SecuredCommand.execute(SecuredCommand.java:87)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.executeCmd(Closure.java:480)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.executeStatement(Closure.java:406)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Pipe.run(Pipe.java:108)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.execute(Closure.java:182)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.Closure.execute(Closure.java:119)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.felix.gogo.runtime.CommandSessionImpl.execute(CommandSessionImpl.java:94)[80:org.apache.karaf.shell.core:4.0.7]
	at org.apache.karaf.shell.impl.console.ConsoleSessionImpl.run(ConsoleSessionImpl.java:274)[80:org.apache.karaf.shell.core:4.0.7]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_102]
```

**Comment by kaikreuzer** [comment]
I would claim that this exception is not directly related to this PR as all I do is to call `playNotificationSoundURI`...

**Comment by maggu2810** [comment]
I think it is correct that is it not related to this PR but this PR shows that there is something wrong in some cases.
I need to add that
* I plugged off the Sonos box from the power supply
* Add the power supply again
* Wait until ESH discovered the box
* Add the thing
* Tested the audio stream

Should I create separate issues for the two exceptions?

This feature shows that there are different problems in the current code base.
Do you agree that new contributions should work in a reliable way?
IMHO we should wait for the merge until the function could be used with a minor change to fail.

I think we should at least fix
* the IP of the local interface detection
* don't rely on the system property for the port
* fix getNotificationSoundVolume NPE
before

Fixes that could be done after merge:
* waitForFinishedNotification NPE
because the audio stream is played correctly

**Comment by kaikreuzer** [comment]
I have fixed the two NPEs.

Interface detection and port handling are treated in other PRs, so I think this here is independent of it.

**Pull request merged**